src/raft_server.c

```diff
@@ -4915,131 +4915,188 @@ raft_server_net_client_request_init_sm_apply(
     ri->ri_last_applied_cumulative_crc = last_applied_cumulative_crc;
 
+
     DBG_RAFT_INSTANCE(LL_TRACE, ri, "");
 }
 
 static raft_server_epoll_sm_apply_t
 raft_server_last_applied_increment(struct raft_instance *ri,
                                    const struct raft_entry_header *reh)
 {
-    NIOVA_ASSERT(ri && reh &&
-                 (reh->reh_index == (ri->ri_last_applied_idx + 1)));
+    NIOVA_ASSERT(ri && reh && 
+        (reh->reh_num_entries > 0));
+    
+    /*
+    Increment the sub apply idx, if its the first sub entry then increment
+    the main last applied idx and update the cumulative crc. If the sub idx
+    reaches the max sub idx then reset both sub idx and max sub idx to -1.
+    */
+    ri->ri_last_applied_sub_idx++;
+
+    if (ri->ri_last_applied_sub_idx == 0) {
+        ri->ri_last_applied_idx++;
+        ri->ri_last_applied_cumulative_crc ^= reh->reh_crc;
+        ri->ri_last_applied_sub_idx_max = reh->reh_num_entries - 1;
+    }
+
+    if (ri->ri_last_applied_sub_idx == ri->ri_last_applied_sub_idx_max) {
+        ri->ri_last_applied_sub_idx = -1;
+        ri->ri_last_applied_sub_idx_max = -1;
+    }
 
-    ri->ri_last_applied_idx++;
-    ri->ri_last_applied_cumulative_crc ^= reh->reh_crc;
+    DBG_RAFT_INSTANCE(LL_NOTIFY, ri, "idx=%ld crc=%x sub-idx=%ld/%ld",
+                      ri->ri_last_applied_idx, reh->reh_crc,
+                      ri->ri_last_applied_sub_idx, ri->ri_last_applied_sub_idx_max);
+}
 
-    DBG_RAFT_INSTANCE(LL_NOTIFY, ri, "idx=%ld crc=%x",
-                      ri->ri_last_applied_idx, reh->reh_crc);
+static bool
+raft_server_needs_apply(struct raft_instance *ri) {
+    NIOVA_ASSERT(ri);
+    NIOVA_ASSERT(ri->ri_last_applied_idx <= ri->ri_commit_idx);
+
+    return (ri->ri_last_applied_idx < ri->ri_commit_idx ||
+            (ri->ri_last_applied_idx == ri->ri_commit_idx &&
+             ri->ri_last_applied_sub_idx < ri->ri_last_applied_sub_idx_max));
+}
+
+static raft_entry_idx_t raft_server_get_to_be_applied_idx(struct raft_instance *ri) {
+    NIOVA_ASSERT(ri);
+    return ri->ri_last_applied_idx + (ri->ri_last_applied_sub_idx == -1 ? 1 : 0);
 }
 
 static raft_server_epoll_sm_apply_bool_t
 raft_server_state_machine_apply(struct raft_instance *ri)
 {
     NIOVA_ASSERT(ri);
-    NIOVA_ASSERT(ri->ri_last_applied_idx <= ri->ri_commit_idx);
 
     DBG_RAFT_INSTANCE(LL_NOTIFY, ri, "");
-
-    if (FAULT_INJECT(raft_server_bypass_sm_apply) ||
-        ri->ri_last_applied_idx == ri->ri_commit_idx)
+    
+    if (!raft_server_needs_apply(ri))
+        return;
+    
+    if (FAULT_INJECT(raft_server_bypass_sm_apply))
         return;
 
-    //reply size for write should be small
-    const size_t reply_buf_sz = RAFT_BS_SMALL_SZ;
-
-    const raft_entry_idx_t apply_idx = ri->ri_last_applied_idx + 1;
-
+    // Read the raft entry header for the to be applied index
+    const raft_entry_idx_t to_apply_idx = raft_server_get_to_be_applied_idx(ri);
```

Comment:
Do we need to say "to_be" here?  The reason I ask is because we already have "last_applied" which signifies that the those indices have already been processed and stored.  So is it not clear enough to simply say:
`apply_idx = raft_server_next_apply_idx(struct last_applied_stuff)` 
using the present tense to imply that it's the entry that we're currently working on?